src/sync.js

```diff
@@ -1246,16 +1253,17 @@ SyncApi.prototype._pokeKeepAlive = function() {
             // responses fail, this will mean we don't hammer in a loop.
             self._keepAliveTimer = setTimeout(success, 2000);
         } else {
+            connDidFail = true;
```

Comment:
Anything but a 400 would trigger this, no? Is that a safe assumption (e.g. 404, 500, ...)?
Rather than doing something like this:
https://stackoverflow.com/questions/20231075/how-to-catch-networkerror-in-javascript

Reply:
True, although given we're polling the /versions endpoint, it could be assumed that it's not going to return anything other than 200 or possibly 400 in the case of a very old HS. We could also have HTTP level errors from HTTP proxies which should probably be classed as connectivity errors. Not sure though - wdyt?

Reply:
Hmm, I'd add `404` to line 1241 because `400` seems more like a synapse idiosyncrasy than anything else?

Reply:
wfm